keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/DoubleVertex.java

```diff
@@ -96,8 +96,38 @@ public MultiplicationVertex multiply(DoubleVertex that) {
         return new MultiplicationVertex(this, that);
     }
 
-    public MatrixMultiplicationVertex matrixMultiply(DoubleVertex that) {
-        return new MatrixMultiplicationVertex(this, that);
+    /**
+     * Matrix product of two vertices
+     *
+     * @param that  a double vertex representing a matrix or a vector to matrix multiply
+     * @return a vertex that represents the matrix multiplication of two vertices.
+     * - If both left and right operands are rank 1, they are promoted to a matrix by prepending a 1 to its dimensions.
+     *   After matrix multiplication, it is reshaped to be a scalar. This is essentially a dot product.
+     *   This returns a ReshapeVertex.
+     * - If only one of the operands is rank 1 (and the other operand is rank 2), it is promoted to a matrix by prepending a 1 to its dimensions.
+     *   After matrix multiplication, the appended 1 is removed. This is essentially a matrix-vector product.
+     *   This returns a ReshapeVertex.
+     * - Otherwise, they are multiplied like conventional matrices.
+     *   This returns a MatrixMultiplicationVertex.
+     */
+    public DoubleVertex matrixMultiply(DoubleVertex that) {
```

Comment:
I wonder if it would be better to return a `ReshapeVertex` under all circumstances? That way you don't need all the dodgy casts to `MatrixMultiplicationVertex` in your tests.

(What you really care about is that you get back a `DoubleVertex` that implements `Differentiable`)

Reply:
Hmmm, I didn't mind it too much because the stuff we are doing in those tests aren't supposed to be used directly by the average user.

Reply:
ok, let's leave it

Reply:
The reshape vertex might have a performance impact if added where it's not actually needed. Although, the performance impact might be worth avoiding the extra code complexity.